src/cargo/ops/cargo_rustc/output_depinfo.rs

```diff
@@ -0,0 +1,128 @@
+use std::borrow::ToOwned;
+use std::collections::HashSet;
+use std::io::{Read, Write};
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use core::{TargetKind};
+use ops::{Context, Unit};
+use util::{CargoResult, internal, human};
+
+#[derive(Debug)]
+struct DepLine {
+    target: String,
+    deps: Vec<String>,
+}
+
+struct DepFile {
+    dir: String,
+    deps: Vec<DepLine>,
+}
+
+fn render_filename<P: AsRef<Path>>(path: P, basedir: Option<&str>) -> CargoResult<String> {
+    let path = path.as_ref();
+    let relpath = match basedir {
+        None => path,
+        Some(base) => match path.strip_prefix(base) {
+            Ok(relpath) => relpath,
+            _ => path,
+        }
+    };
+    relpath.to_str().ok_or(internal("path not utf-8")).map(ToOwned::to_owned)
+}
+
+fn read_dep_file<P: AsRef<Path>>(path: P) -> CargoResult<DepFile> {
+    let mut file = File::open(&path).map_err(|_|
+        human("error opening ".to_string() + path.as_ref().to_str().unwrap_or("(bad unicode")))?;
+    let mut contents = String::new();
+    let _ = file.read_to_string(&mut contents)?;
+    let mut spl = contents.split('\0');
+    let dir = spl.next().ok_or(internal("dependency file empty"))?;
+    let dep_txt = spl.next().ok_or(internal("dependency file missing null byte"))?;
+    let mut result = Vec::new();
+    for line in dep_txt.lines() {
+        let mut line_spl = line.split(": ");
+        if let Some(target) = line_spl.next() {
+            if let Some(deps) = line_spl.next() {
+                let deps = deps.split_whitespace().map(ToOwned::to_owned).collect();
+                result.push(DepLine {
+                    target: target.to_string(),
+                    deps: deps,
+                });
+            }
+        }
+    }
+    Ok(DepFile {
+        dir: dir.to_string(),
+        deps: result,
+    })
+}
+
+fn add_deps(depfile: &DepFile, deps: &mut HashSet<PathBuf>) {
+    let dep_dir = PathBuf::from(&depfile.dir);
+    for depline in &depfile.deps {
+        for dep in &depline.deps {
+            deps.insert(dep_dir.join(dep));
+        }
+    }
+}
+
+// TODO: probably better to use Context::target_filenames for this
```

Comment:
Yeah I think we'll just want to emit one `.d` file with the same contents for each filename returned from `target_filenames`